### PR TITLE
Update --save-preds to save metric value to log and tidy up naming.

### PR DESCRIPTION
**Pull request by am2145**
## Description
This PR adds functionality for the `--save-preds` argument in `train.py` to also compute the model metrics and record these values in the log file. This is to have easy access to performance when running the v2 benchmarks. 

As a note: The metrics computed when calling `trainer.test()` compute them per batch and then perform a reduction (by default mean reduction) across the batches at the "end of epoch" according to lightning's documentation [here](https://lightning.ai/docs/pytorch/stable/extensions/logging.html#automatic-logging). This gives the same value as if running `trainer.predict` and using the predictions to compute the metric across the whole test set for certain metrics like MAE and MSE (subject to @KnathanM's #786 being merged in), but will return different values for other metrics like RMSE for example. 

To distinguish between these two metrics, I have modified the name of the metrics returned by `trainer.test()` to be called 'batch averaged test' and called the name of the metrics computed when calling `--save-preds` to be called  "Entire Test Set" to distinguish between the two so that it's clear they can be different values.

### CODE REVIEW

**Comment by shihchengli** [comment]
@am2145 It seems you accidentally added some .toml files in [a766699](https://github.com/chemprop/chemprop/pull/788/commits/a766699f66d2873d9f5066141b372183adf00dbb), which led to the tests failing.

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -824,11 +825,39 @@ def train_model(args, train_loader, val_loader, test_loader, output_dir, scaler,
...
+                targets = test_dset.Y
+                mask = torch.from_numpy(np.isfinite(targets))
+                targets = np.nan_to_num(targets, nan=0.0)
+                w_s = torch.from_numpy(test_dset.weights)
+                w_t = model.w_t
+                lt_mask = torch.from_numpy(test_dset.lt_mask) if test_dset.lt_mask[0] else None
+                gt_mask = torch.from_numpy(test_dset.gt_mask) if test_dset.gt_mask[0] else None
+                preds_losses = [
+                    metric(
+                        torch.from_numpy(preds),
```
Comment (KnathanM) [inline chemprop/cli/train.py:846]: Looks like you above you make press a numpy array but don't do any numpy operations on it. Could you do preds = torch.concat(preds, 0) and then not convert to torch here?

File: chemprop/cli/train.py [hunk 2]
```diff
@@ -824,11 +825,39 @@ def train_model(args, train_loader, val_loader, test_loader, output_dir, scaler,
...
+                if isinstance(test_loader.dataset, MulticomponentDataset):
+                    test_dset = test_loader.dataset.datasets[0]
+                else:
+                    test_dset = test_loader.dataset
+                targets = test_dset.Y
+                mask = torch.from_numpy(np.isfinite(targets))
+                targets = np.nan_to_num(targets, nan=0.0)
+                w_s = torch.from_numpy(test_dset.weights)
+                w_t = model.w_t
+                lt_mask = torch.from_numpy(test_dset.lt_mask) if test_dset.lt_mask[0] else None
```
Comment (KnathanM) [inline chemprop/cli/train.py:842]: I think you need to check if lt_mask[0] is not None. If it is False and the rest are true, the whole lt mask will be set to None. Same with gt mask.

File: chemprop/cli/train.py (hunk 1 above)
Comment (am2145) [reply chemprop/cli/train.py:846]: I need it in numpy array form when saving the dataframe to CSV, as torch tensors do not save nicely. However, I could not convert it to numpy until right before I save the predictions. Would that be preferable?

File: chemprop/cli/train.py (hunk 2 above)
Comment (am2145) [reply chemprop/cli/train.py]: Nice catch. I have made this change.

**Review by davidegraff — COMMENTED** [review]
why don't we just simplify the workflow to _always_ just make predictions on the test set and then calculate the loss of the entire prediction tensor? Even in the case of ultra-large test sets ($N \simeq 1 \cdot 10^6$), this shouldn't be problem for memory and it would allow us to simplify the logic and get rid of the `trainer.test()` call.

**Comment by am2145** [comment]
> why don't we just simplify the workflow to _always_ just make predictions on the test set and then calculate the loss of the entire prediction tensor? Even in the case of ultra-large test sets (N≃1⋅106), this shouldn't be problem for memory and it would allow us to simplify the logic and get rid of the `trainer.test()` call.

I would be fine with removing the call to `trainer.test` and moving the calls to `trainer.predict` and metric computation in `--save-preds` to always be done by default. All `--save-preds` would do is then actually write the predictions to a CSV file. I guess an open question is when comparing to other results in the literature, are most other people using `trainer.test` and comparing these batch-averaged metrics? That could be a reason to keep `trainer.test`...

**Comment by KnathanM** [comment]
As a test I used `trainer.test` on a test set of ~40,000 molecules and it took 7 seconds. `trainer.predict` took 2 minutes, I think because it uses a batch size of 1. So it is slower, but in the bigger picture I don't think 2 minutes is a problem. So skipping .test and only using .predict sounds fine to me.

**Comment by KnathanM** [comment]
I'm not sure myself. I added a print statement to `predict_step` of model` and saw that the batches are always a single molecule and always in the order of the dataset even if the dataloader has shuffle=True. The output of predict is a list of 1 x n_task tensors. A difference between predict and test is that in predict all the predictions are saved while in test they are discarded. Maybe instead of returning a list of batched predictions, the lightning people though a list of single predictions was better. I also don't see this in the documentation.

**Comment by davidegraff** [comment]
That's odd- I've never encountered that issue before. In my hands, the output of `trainer.predict()` is `list[T_out]`, where `T_out` is the return type of `predict_step()`, in this case a tensor of shape $(B,\ T)$, $B$ is the batch size and $T$ is the number of output tasks. You can then inline it like so:

```python
Y_hat = torch.cat(trainer.predict(model, test_loader), 0)
```

There must be something odd going where the batch size of the `test_loader` is not getting set properly.

**Comment by KnathanM** [comment]
I checked again, and with Chemprop I get `trainer.test(mpnn, test_loader)` uses the correct batch size while `trainer.predict(mpnn, test_loader)` uses batch size of 1. But if I make a native dataset using DataLoader, TensorDataset and a simple model based on LightningModule, then trainer.predict does use batch size not 1. So it is something on the chemprop end. I'll keep looking.

**Comment by KnathanM** [comment]
I found the odd thing, see #808

**Comment by am2145** [comment]
I updated `train.py` to always use `trainer.predict()` to compute metrics on the full test set as per @davidegraff 's suggestion . I removed the call to `trainer.test()` accordingly. Additionally, as per discussion with @KnathanM, since we are always running `trainer.predict()`, I also modified the default behaviour to always save the predictions to the CSV and remove `--save-preds` as an argument. If we want to retain/change the original behaviour, then let me know.

**Review by KnathanM — APPROVED** [review]
Two notes, 
this adds another place #779 will need to rename w_t to task weights
this uses a print a print statement for the metric on the test set. This is because we are replacing trainer.test with trainer.predict. trainer.test would print the metric value to stdout regardless of the `--log` and `-v` flags. using print mimics this behavior.
